add spec for #1620

SirusCodes opened the pull request:
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?

-----

Create a spec as requested by @denelon at https://github.com/microsoft/winget-cli/issues/1620#issuecomment-1206565768 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2650)

SirusCodes committed:
add spec for #1620

SirusCodes commented:
@microsoft-github-policy-service agree

SirusCodes commented:
@denelon can you review it?

denelon commented:
My "meta" feedback is this seems to work well for installing a single package. 

We shouldn't have ambiguity with import since the source and the ID need to be specified in the packages.json file.

I wonder if we should have a "default" behavior and settings to change the default behavior. It may not be necessary, but this is a departure from the current behavior. 

I'm also not sure what this is going to "sound" like with narrator running.

Our telemetry needs to be reasoned about as well. If a user runs `winget install <something>` and we're not able to narrow it down to a single package, I don't want to have some "false" increased concept of "success".

denelon commented:
@yao-msft what are your thoughts here?

yao-msft commented:
This package selection mechanism could be extended to `winget upgrade <single package>`, `winget uninstall <single package>` and `winget show <single package>` as well, since these use the same worflow to pick a package internally.

Regarding default behavior, I think it'd be better to make this one opt-in from a specific argument since most automated systems, script writers will not be compatible with this change. Or we should announce this "breaking" default behavior well ahead before we make this the default behavior. In the early winget design principal, winget tries not to prompt for user input unless it has to(i.e. accepting agreements, missing arguments, etc), and the prompt can be disabled in many ways.

Regarding installing multiple packages in one command, it requires further thinking and framework changes are likely required to support this. Command line syntax may also need to be revisited.

denelon commented:
@yao-msft 
So maybe something like adding "--prompt" could prompt if the results are ambiguous, and a setting could be added to make the setting default and "--no-prompt" could override the setting via the CLI?

yao-msft commented:
> @yao-msft So maybe something like adding "--prompt" could prompt if the results are ambiguous, and a setting could be added to make the setting default and "--no-prompt" could override the setting via the CLI?

We already have a cli arg and user setting both called "DisableInteractivity" to indicate "no prompt", it's opted in for com callers(can not be overriden) but not from cli. So I guess the current default in cli is already "allow prompt". But currently only very few packages will trigger a prompt. With this feature, we'll see prompts a lot more, so I'm wondering if we should have a specific opt in for this(from settings), we may change the default behavior in future. In another thought, if existing scripts work, they should already be able to point to a specific package, so it might not be that "breaking" for existing scripts and automated systems. I think a setting for opt in is enough. And I'm open to other suggestions as well.

SirusCodes commented:
> This package selection mechanism could be extended to `winget upgrade <single package>`, `winget uninstall <single package>` and `winget show <single package>` as well, since these use the same worflow to pick a package internally.

Yes this can/should be extended to those as well. Will make that change in the spec. 

> Regarding default behavior, I think it'd be better to make this one opt-in from a specific argument (or settings?) since most automated systems, script writers will not be compatible with this change. Or we should announce this "breaking" default behavior well ahead before we make this the default behavior. In the early winget design principal, winget tries not to prompt for user input unless it has to(i.e. accepting agreements, missing arguments, etc), and the prompt can be disabled in many ways.

Can you please elaborate on this more? As I feel it will only come in if there are multiple matches and we get a table of packages in current implementation.
Everyone would write a script using Id's which are a unique to each package and if they get the selector they will get the current table in current implementation.
Only difference would be that in current implementation the command would return and not wait for an user input.

> Regarding installing multiple packages in one command, it requires further thinking and framework changes are likely required to support this. Command line syntax may also need to be revisited. 

Yeah, that is not the goal of this spec. I added that because if in future we plan to add that this is how selector should work.

SirusCodes commented:
> With this feature, we'll see prompts a lot more

How? This will only come into picture if there are multiple matches in current implementation.

For example.

```pwsh
winget install Powershell
```

It currently gives you 2 packages and after this it will also give you 2 packages just will a easy selector.

SirusCodes commented:
> I'm also not sure what this is going to "sound" like with narrator running.

Yeah I'm also not sure about that but if it reads it line by line then there shouldn't be a big problem.
Maybe I can try implementing a dummy one and see.

> Our telemetry needs to be reasoned about as well. If a user runs `winget install <something>` and we're not able to narrow it down to a single package, I don't want to have some "false" increased concept of "success".

I haven't thought about this. Can you help me out with it?

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+The above command will not install anything.
+
+If will ever support this then we might consider having _multiple_, _range_ and _not_ selectors.
```
```suggestion
If this is ever supported, potential considerations include having _multiple_, _range_ and _not_ selectors.
```
It is best practice to not reference a specific viewpoint such as `we`

SirusCodes replied:
Thanks, will keep this in mind.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
Will the user be required to press enter to confirm their input?

SirusCodes replied:
Yes

Trenly replied:
That should be included in the spec then

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
```
There is a potential issue where non-interactivity would need to be preserved. Please add in specifications for how the client should behave when interactivity is disabled through the COM context, through the `--disable-interactivity` argument, or through user settings.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
What is the UI/UX if the user provides invalid input?

SirusCodes replied:
We can tell them this is an invalid input. Please give the correct input.

Trenly replied:
That error message should be included in the spec then

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
What rules constitute valid input?

SirusCodes replied:
The packages will be placed in the listed in a table. User will type the row number

Trenly replied:
So if a user entered `1` it would be valid, but what if they entered `1.` or `1withSomeText` or `99999`. These need to be included in the spec. Something like `Only entries which exactly match a row number are considered valid input` with an example table output and then bullet point examples of what would be considered valid or invalid

SirusCodes replied:
Yeah will add that in docs

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
How does the user exit the selector if they don't want to choose any of the available options?

SirusCodes replied:
We can have something like "0 to exit"

Trenly replied:
That should be included in the spec then

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+## Capabilities
+
+### Accessibility
+
```
New text in the menu (e.g `Please select the packages`) will need to be localized. There may need to be considerations made for if the indices used for selecting packages also need to be localized

SirusCodes replied:
Do windows localize numerical forms automatically?

Trenly replied:
I'm not sure, but it would need to be a consideration that needs to be made as to whether they will or won't be

SirusCodes replied:
I will verify it and add in the spec

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
Is there a limit to how many items can be in the selector?

SirusCodes replied:
Haven't thought about it. But my initial thought is to show all the matches which we get🤔

Trenly replied:
If I run `winget install -q " "` there would be 1899 entries for the user to select from, and that's just using the winget source, not including the msstore source or private sources. I'm not saying that it's necessarily a bad idea to show them all, since the current behavior of `winget search` is to show them all.

I believe that once you exceed 10-20 options, the user input wasn't good enough to begin with and it should revert to the list behavior without the selector, with a message like 'please refine your query`

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
Is tab completion valid for the selector?

SirusCodes replied:
No

Trenly replied:
Why wouldn't it be? It could very easily cycle through the row indices. I'd rather hit <kbd>Tab</kbd> <kbd>Enter</kbd> than <kbd>1</kbd> <kbd>Enter</kbd> or <kbd>Tab</kbd> <kbd>Tab</kbd> <kbd>Enter</kbd>than <kbd>2</kbd> <kbd>Enter</kbd>

Either way, whether it is supported or not supported should be in the spec

SirusCodes replied:
I guess it's a personal choice. I prefer `1` `Enter`.
Surely, it can support both, but should it?

What are your opinions @denelon @yao-msft?

denelon replied:
I'm curious as to whether I would prefer to enter the "number" or the "ID" to disambiguate. My initial thought is that I would type the number.

The number is certainly less typing, but when we think about the tab completion it would likely be less error prone in either case.

If we support both a number or a package identifier, I think I would want the tab to prioritize the package ID first just for the sake of clarity.

In that case, the tab auto complete would be pretty fantastic.

Trenly replied:
Interesting thoughts;  Although that does bring up the edge case where there may be two packages with the same Identifier from different sources, in which case the number would have to be used instead of the package identifier.

Perhaps that calls for some sort of logic like -
> If user enters a row number, treat it as a selection. 
> If the user has entered text, treat it as a new query using only the table results as the source

That would mean that the behavior would be as follows, assuming a table like this:
```pwsh
> winget install Example
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example         Example.Package custom
 2  Example Beta    Example.Package winget
 3  Example Browser Example.Browser winget

Please enter a row number or refine your query.
> _
```

User enters `2` -
```pwsh
> 2
Found Example Beta [Example.Package] *****
Downloading ******
```
User enters `1` -
```pwsh
> 1
Found Example [Example.Package] *****
Downloading ******
```
User enters `Example.Browser` -
```pwsh
> Example.Browser
Found Example Browser [Example.Browser] *****
Downloading ******
```
User enters `--source custom`
```pwsh
> --source custom
Found Example [Example.Package] *****
Downloading ******
```
User enters `--source winget` -
```pwsh
> --source winget
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example Beta    Example.Package winget
 2  Example Browser Example.Browser winget

Please enter a row number or refine your query.
> _
```
User enters `Example.Package` -
```pwsh
> Example.Package
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example         Example.Package custom
 2  Example Beta    Example.Package winget

Please enter a row number or refine your query.
> _
```
User enters `--source msstore` -
```pwsh
> --source msstore
No package found matching selection criteria.
```

SirusCodes replied:
Thanks @Trenly this looks pretty powerful

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```
Are users required to select an index, or could they enter a package ID if they wanted to?

SirusCodes replied:
Both seems fine to me, but index is easier in my opinion. What do you feel?

Trenly replied:
I'd be more inclined to flexibility. As @denelon likes to say - meet the users where they are. As a user, I would feel more confident entering a full package ID just to have the peace of mind that I didn't make a typo, but I can also see where others would want to use the index

SirusCodes replied:
Yeah, supporting both seems like a better idea. Will add it in the doc.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+### Reliability
+
+This will help the users to have a better experience and would decrease the chance of human error while copy-pasting the Id.
```
Will there need to be any measures taken against mistyping the selection? It is a lot easier to accidentally type `2` instead of `3`  than it is to enter an incorrect package ID.

SirusCodes replied:
Yeah, it's something which is something that can happen. Adding another prompt to take the confirmation saying "Do you want to install XYZ package? (Y/n): ". But again, it's an extra step.

I would add this in the in doc. Don't have any idea how we can handle it

Trenly replied:
Perhaps it's a future consideration

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+## Inspiration
+
+It is not a good experience to copy Id and then run the command again.
```
Some users (me especially) will not want to use this feature. There should be a way to toggle this feature off in settings, outside of the setting for disabling interactivity.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
```
There is potential confusion when using the `--silent` argument. As a user, I would expect that `--silent` would not lead to an interactive menu. Having reviewed the spec and being familiar with how Winget works, I understand that the selector is not part of the package install, but this could be a sticking point for many new users and should be preempted.

SirusCodes replied:
Yeah, we should not show selector when `--silent` argument is passed.

Trenly replied:
> Yeah, we should not show selector when `--silent` argument is passed.

I think more opinions would be needed here. IMO `--silent` shouldn't control how winget behaves as a CLI, that's what the `--disable-interactivity` prompt is for. My point was more that some mention should be made around this, or it should be documented that `--silent` doesn't suppress this and that `--disable-interactivity` would. @yao-msft and @ryfu-msft would be the experts here though

Trenly commented:
> I wonder if we should have a "default" behavior and settings to change the default behavior. It may not be necessary, but this is a departure from the current behavior.

I think this is a necessity, especially for this feature. As was mentioned in the original issue and I touched upon in my review, there are users who will not want this feature at all - even outside of a scripting perspective. I would be a strong advocate for including user settings.

> Our telemetry needs to be reasoned about as well. If a user runs winget install <something> and we're not able to narrow it down to a single package, I don't want to have some "false" increased concept of "success".

I think this is as "simple" as adding a context flag that gets set whenever the interactive selector gets called, and sending that flag along with the telemetry. I would still count it as a success if the user was able to manually disambiguate, but I can understand if the internal needs differ here.

> This package selection mechanism could be extended to winget upgrade <single package>, winget uninstall <single package> and winget show <single package> as well, since these use the same workflow to pick a package internally.

I think implementing this for upgrade would conflict with [#2627](https://github.com/microsoft/winget-cli/issues/2627). It may be a good feature in the meantime though. I agree with adding it for uninstall and show.

> With this feature, we'll see prompts a lot more, so I'm wondering if we should have a specific opt in for this(from settings), we may change the default behavior in future. In another thought, if existing scripts work, they should already be able to point to a specific package, so it might not be that "breaking" for existing scripts and automated systems. I think a setting for opt in is enough. And I'm open to other suggestions as well.

I go both ways on this. On one hand you have people like me who are technically inclined that would prefer to see and work with CLI output directly rather than have any form of interactivity in a CLI application, except that which is strictly necessary. On the other hand you have people who are using winget to install their favorite application and want it to be as quick and seamless as possible, and having interactivity would help. I think opt-out is easiest to implement from the `DisableInteractivity` perspective, and keeps the same current default behavior as other interactivity, but I prefer the original design principal of not asking for input unless we have to. I say that `opt-in` would be my final preference. (Apologies for long section, had to write my thoughts down for me to understand them myself)

> Regarding default behavior, I think it'd be better to make this one opt-in from a specific argument (or settings?) since most automated systems, script writers will not be compatible with this change

I agree. Not only because of the points mentioned above, but I know that some scripts like [marinet101/WingetUI](https://github.com/martinet101/WingetUI/blob/53222c6baeafd9a1669c8fbe089120d78f3be24b/wingetui/wingetHelpers.py#L174) actually use the table output and parse the text. Adding in an additional column here or making it so that the feature is opt-out would be a breaking change for them. 

> Everyone would write a script using Id's which are a unique to each package and if they get the selector they will get the current table in current implementation.

Id's aren't always guaranteed to be unique to a package. The item which is guaranteed to be unique is the Source-ID pair. It is entirely possible that two sources have the same identifier for different packages. 

> In another thought, if existing scripts work, they should already be able to point to a specific package, so it might not be that "breaking" for existing scripts and automated systems.

The more I think about this, I believe that the dynamic nature of sources would make this a breaking change. It is entirely possible that a script which is working today could be broken tomorrow by someone adding a new package to the repository, assuming the script was pointing at a volatile field like a package name; Or even in cases where a sub-package is added and there is potential for a moniker conflict that we aren't aware of. I would suspect that a sysadmin would rather have 1 package that failed to install with a very clear log about why as opposed to coming back several hours later just to find that the script hung because they weren't aware of the `--disable-interactivity` flag, and thought there would be no issues since the script was working fine before.

SirusCodes commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,93 @@
+
+### Compatibility
+
+It will not break anything.
```
Oh, never knew this. Thanks for highlighting this.

SirusCodes committed:
Apply suggestions from code review

Co-authored-by: Kaleb Luedtke <trenlymc@gmail.com>

denelon commented:
I think the default behavior should not change. This would certainly be breaking. I would expect to have a setting for "disambiguation". maybe "prompt" and "no-prompt". Then if one had the default they could override with "--prompt" or "--no-prompt".

The default setting would be "no-prompt"

Overriding the default setting would look like:

`winget install PowerToys --prompt`

Changing the default setting would look like:
```JSON
"packageDisambiguation": {"prompt"}
```
Overriding the "prompt" setting would look like:

`winget install PowerToys --no-prompt`

denelon commented:
Changing this behavior to be the default would require:
* #2128

SirusCodes committed:
improve spec based on suggestions

SirusCodes commented:
Hey @denelon, @yao-msft and @Trenly for your suggestions. I have updated the specs from suggestions you have given.

- How it will handle `--disable-interactivity`
- How it will handle invalid inputs
- How user can refine their search results
- Added how the settings should look like
- Added a compatibility issue with MinTTY (#2013) 
- Added a limit on the number of packages shown
- Made it an opt-in feature



Tho there are a few things where I don't have much knowledge about and would be great if you can help me out in those parts.

> I think implementing this for upgrade would conflict with https://github.com/microsoft/winget-cli/issues/2627. It may be a good feature in the meantime though. I agree with adding it for uninstall and show.

@Trenly I'm not sure how this will conflict it 

---

> I think this is as "simple" as adding a context flag that gets set whenever the interactive selector gets called, and sending that flag along with the telemetry. I would still count it as a success if the user was able to manually disambiguate, but I can understand if the internal needs differ here.

I don't have any knowledge how telemetry and localization work and hence couldn't add anything about it in spec.

It would be great if you can point be where I can get to know about those.

---

I tried by printing an example on terminal with screen reader on. It read me the whole thing correctly but read `No.` (abbreviation of number) as `No` (as in answer to question).

Is there a way where we can get to know that the screen reader is on and print `Number` instead of `No.` or we should stick with `Number`?

--- 

I need your suggestions on how refine search should work. Should it refine results on the basis of exact match or match pattern?

I'm inclined towards matching patterns, but it may result into extra steps for users. Like if there are 2 packages with name Example and a package with name Example2 pattern matching will result in to retaining all three if user types Example to refine them. But at the same time, it will help users to type less to refine results.

Again, I would like it to be as pattern matching thing instead of exact match.

---

As previously discussed in [review comment](https://github.com/microsoft/winget-cli/pull/2650#discussion_r1010012165) to have tab selection.

I'm not sure if it will be intuitive for users to use `Tab` to cycle through number. But if we are going with this, I want your opinion on if it should only cycle through number or package names as well? It doesn't make much sense to me to have both, cycling through numbers is more than enough in my opinion.

Also, if we are having `Tab` to cycle forward then we should have `Tab` + `Shift` to cycle backwards.

Trenly commented:
> @Trenly I'm not sure how this will conflict it

If a user were to type `winget upgrade Microsoft` - there are multiple packages that match. The conflict would be whether the selection menu is shown, or if all packages matching `Microsoft` would be upgraded

> Is there a way where we can get to know that the screen reader is on

I'm certain this is possible, it would just be a matter of how big the change would be to support this. It's probably better to use the full text `Number` all the time, or an alternate name like `Index` 

> I need your suggestions on how refine search should work. Should it refine results on the basis of exact match or match pattern?

I would suggest this be implemented as a temporary source. Essentially, the results in the table (including those that are truncated), would become a local source, and the default query behavior would be used. I believe this is a substring match, but I'd leave it up to whatever the current default search behavior is

>  I want your opinion on if it should only cycle through number or package names as well

My interpretation of denelon's comments is that tab completion would be most beneficial for cycling through Package Identifiers, and wouldn't be as helpful / necessary for numbers.

> Also, if we are having Tab to cycle forward then we should have Tab + Shift to cycle backwards.

I agree. This is the current implementation of the tab completion used in the client, and the behavior should be carried forward.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,235 @@
+
+- Number - It should match with a _row number_ (0 will exit the selector)
+- Text - It should match either the package _Name_ or _Id_
+- `--source` flag - It refines the search with a specific package source. It can only be used if a user enters text or standalone.
```
I would extend this to be any of the default query flags. Essentially, I think of the table results as a new source that can be queried against, and whatever is input is the [[-q] <query>] for a search against the new source.

Probably don't need things like `-n` or `--header` or `--accept-source-agreements`, but I would imagine that a majority of the options would still be available
```
Searches for packages from configured sources.

usage: winget search [[-q] <query>] [<options>]

The following command aliases are available:
  find

The following arguments are available:
  -q,--query                  The query used to search for a package

The following options are available:
  --id                        Filter results by id
  --name                      Filter results by name
  --moniker                   Filter results by moniker
  --tag                       Filter results by tag
  --cmd,--command             Filter results by command
  -s,--source                 Find package using the specified source
  -n,--count                  Show no more than specified number of results (between 1 and 1000)
  -e,--exact                  Find package using exact match
  --header                    Optional Windows-Package-Manager REST source HTTP header
  --accept-source-agreements  Accept all source agreements during source operations
```

SirusCodes replied:
I never thought it as CLI inside a CLI. Wouldn't that be an overkill for a simple thing? 👀

Trenly replied:
> I never thought it as CLI inside a CLI. Wouldn't that be an overkill for a simple thing? 👀

It isn’t a CLI in a CLI though.

I'll use the example of the default `winget` source. That source is a pre-indexed source, which is built from the community repository. At its core, the `winget` source is just a database file, called `index.db`. When a user runs `winget search`, the CLI runs several queries against the index (index.db) to come up with a list of the packages matching the search criteria. 

My suggestion is, take all of the results that are returned and store them in a temporary database file. Then, a user can use *any* of the default search parameters when refining  the query.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,235 @@
+
+### Number of results
+
+We should add a `--list-count` flag to customize the number of packages listed in the prompt. By default, the user will see a list of 20 packages.
```
Users already have the `-n,--number` flag they can pass; Perhaps this would be better as a user setting?

```json
 "Interactivity": {
      "description": "Interactivity settings",
      "type": "object",
      "properties": {
        "disable": {
          "description": "Controls whether interactive prompts are shown by the Windows Package Manager client",
          "type": "boolean",
          "default": false
        },
        "listItemLimit":{
	  "description": "Controls how many items are shown in interactive selection lists",
	  "type": "integer",
          "default": 20,
          "minimum": 0
        }
      }
    },

SirusCodes replied:
Oh, we can use that... we don't need to add more I guess

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,235 @@
+
+We should add a `--list-count` flag to customize the number of packages listed in the prompt. By default, the user will see a list of 20 packages.
+
+If the number of packages is more than allowed it should show a text saying how many more packages are available
```
How would users see the additional packages if they don't want any of the first 20?

SirusCodes replied:
I thought about it. I felt it was not necessary but more I think about it I feel it's good to have.

`-m` and `--more` flag is ok I feel.

Will add it in the next iteration.

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
````diff
@@ -0,0 +1,235 @@
+Downloading Example (Example.Example2)...
+```
+
+If the `--source` query matches nothing then ask the user to retry.
````
I think this gets back to the idea of preference vs requirement. If the user specifies the source, it would be a requirement and not a preference, so if no packages are found using the source, then I would expect the result to be `No package found matching input criteria` and the menu to exit.

> A short note on preferences vs requirements. The original concept has always been that there are user preferences, and user requirements. User requirements must be met in order to proceed; example - a user sets a requirement of x64 packages only, an x86 package can never be installed. If a manifest only contains x86 versions, it won't be available to the user. User preferences change the prioritization, but do not need to be met to proceed; example - a user sets a preference of en-GB for Installer Locale. If a manifest contains the en-GB locale, it will be used, regardless of what the system locale is, but if the en-GB locale is not present, the installation will continue with the most applicable locale.
>  
> User requirements and preferences can be set through the settings file. However, any command line switches override the settings and are treated as a requirement, not as preference. So if the user has a requirement for x64 architecture, but specifically uses `--architecture x86`, then the requirement becomes x86 just for that run of the install command.

SirusCodes replied:
Oh, that makes so much sense. Thanks for that. I always that it as a preference.

SirusCodes commented:
> If a user were to type `winget upgrade Microsoft` - there are multiple packages that match. The conflict would be whether the selection menu is shown, or if all packages matching `Microsoft` would be upgraded

So that should be added in _potential issues_ section, right?

> I'm certain this is possible, it would just be a matter of how big the change would be to support this. It's probably better to use the full text `Number` all the time, or an alternate name like `Index`

`Number` feels a better user-friendly word. 
 
> I would suggest this be implemented as a temporary source. Essentially, the results in the table (including those that are truncated), would become a local source, and the default query behavior would be used. I believe this is a substring match, but I'd leave it up to whatever the current default search behavior is

I feel default one is pattern matching by looking at the `winget upgrade Microsoft` example.
 
> My interpretation of denelon's comments is that tab completion would be most beneficial for cycling through Package Identifiers, and wouldn't be as helpful / necessary for numbers.

---

Perfect, will make the necessary changes in next iteration.

SirusCodes committed:
add more flags
use `-n` flag for result number
use `-m` flag for more results
tab completion
conflict in `upgrade` command
minor fixes

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,270 @@
+
+The user needs to **opt-in** to this feature.
+
+Winget can use `--prompt` and `--no-prompt` flags to enable and disable it respectively.
```
```suggestion
Winget can use `--prompt` and `--disable-interactivity` flags to enable and disable it respectively.
```

SirusCodes replied:
I used `--no-prompt` because @denelon suggested it in https://github.com/microsoft/winget-cli/pull/2650#issuecomment-1297777064

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
````diff
@@ -0,0 +1,270 @@
+
+The user can change the default setting to always show the prompt by -
+
+```pwsh
+"packageDisambiguation": "prompt"
````
```suggestion
```json
"interactivity" " {
  "packageDisambiguationPrompt": true
}
```
It would be best to use the existing settings structure

SirusCodes replied:
This was something from @denelon's https://github.com/microsoft/winget-cli/pull/2650#issuecomment-1302541619

Trenly replied:
It isn’t very clear what this is for then, as the way it is described here doesn’t match denelons comment.

I'll take a try at making some smaller tweaks and will create a PR into your branch

SirusCodes replied:
Merged

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
````diff
@@ -0,0 +1,270 @@
+
+and disable it by -
+
+```pwsh
+"packageDisambiguation": "no-prompt"
````
```suggestion
```json
"interactivity" " {
  "packageDisambiguationPrompt": false
}
```

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,270 @@
+
+### Tab completion
+
+The Tab completion would complete the package name/ID and cycle through the package ID on `Tab` press and reverse cycle through it on `Tab` + `Shift`.
```
```suggestion
The Tab completion would complete the package name/ID and cycle through the package ID on <kbd>Tab</kbd> press and reverse cycle through it on <kbd>Tab</kbd>+<kbd>Shift</kbd>.
```

Trenly commented on doc/specs/#1620 - Installer selector on multiple matches.md:
```diff
@@ -0,0 +1,270 @@
+
+### Invalid inputs
+
+If a user enters an invalid value it should show an error asking the user for input again,
```
I think this could be a point of contention. If using the value as additional refinement, I would prefer it to exit with the message `No package found matching input criteria`. However, this would be a UX choice that denelon or someone else at MSFT would need to make

SirusCodes replied:
@denelon or @yao-msft can you please suggest something here?

denelon replied:
Are you looking for a suggestion on the error or something related to the heuristic for determining error cases?

Trenly replied:
@denelon - Already resolved it with my recent PR into the forked branch

Trenly committed:
Move terminoligy to refinement instead of selector

SirusCodes committed:
Merge pull request #1 from Trenly/Spec1620

Update Spec

SirusCodes commented:
@Trenly should we also rename of file to _Query refinement.md_?

Trenly commented:
> @Trenly should we also rename of file to _Query refinement.md_?

Probably a good idea

SirusCodes committed:
rename spec file name

SirusCodes commented:
@denelon, @yao-msft can you have a look?

yao-msft commented:
> @denelon, @yao-msft can you have a look?

Hi @SirusCodes, 

Thanks for your contribution and continued refinement of the spec. We are currently in the process of getting a release candidate of winget v1.4.0. We will get back to the spec when this feature work is scheduled.

Usually the spec will be checked in around the time implementation is done. And the spec might need to be modified and extended to match the actual implementation. I.e. features actually done, features moved to be considered in the future, a little description on the technical implementation like data schema, and expected behaviors and known issues, etc.

Thanks,
Yao

bidaumm commented:
It appears this has been stagnated since Dec '22
Is this still on the cards for future implementation?

denelon commented:
Yes, we still want to offer an "interactive" mode for WinGet. We've been discussing several scenarios internally. This is one of the topics most likely to lead towards a WinGet 2.0.

Ideally, the default behavior for that version of WinGet will support better interactive experiences for users. We're also very aware of the desire to have non-interactive scripted experiences. We've been discussing the impact for automated scenarios and how we would have the client respond so scripted behaviors could more easily isolate the failures encountered by interruptions caused by interactive prompts expecting a user to respond.

If we offer the interactive experience as an option a user could "opt-in" we would need to figure out the best way to help them discover this new mode. We would certainly need settings and command-line arguments to toggle the behavior as a default mode as well as an override option.

mominshaikhdevs commented:
what's the latest update on this?

Trenly commented on doc/specs/#1620 - Query Refinement for Disambiguation.md:
```diff
@@ -0,0 +1,220 @@
+## Future considerations
+* Keyboard interactivity using <kbd>↑</kbd>, <kbd>↓</kbd>, and <kbd>Ener</kbd>
+
+* Currently, Winget does not support installing multiple packages in a single command.
```
This has changed since the last iteration of the spec;

It should probably be updated so that each package which needs disambiguation goes through the disambiguation prompt at the start of the install flow.

SirusCodes replied:
@Trenly are you talking about https://github.com/microsoft/winget-cli/pull/4282 this? If yes, then what needs to be considered?

Trenly replied:
> @Trenly are you talking about #4282 this? If yes, then what needs to be considered?

No, I'm talking about #2861

SirusCodes commented:
Hi all, I'm a bit busy. Will try to look at the PR this weekend.

SirusCodes committed:
update #1620 spec to consider multiple package installation

SirusCodes commented:
@Trenly @denelon I have updated the spec.

I thought disambiguating each package separately would be a better idea because it will leave out edge cases where a user might filter out the package they wanted to install and go back through this whole process.

denelon commented:
@SiriusCodes, I've done a readthrough. The spec looks fairly thorough to me. At this point I'd like to get one of the engineers to perform a bit of a sanity check 😊

I'm also thinking we need to reason about what this would imply with the PowerShell way of doing things. I'm not sure it makes sense to have this kind of treatment in PowerShell though, so I'll see if I can get a few folks to look at it and give feedback.

@JohnMcPMS

JohnMcPMS commented on doc/specs/#1620 - Query Refinement for Disambiguation.md:
```diff
@@ -0,0 +1,275 @@
+From the command line, users should be able to use `--refine-query` to override the user setting and show the query refinement, and `--no-refinement` to override the user setting and not show the query refinement.
+
+### Refinement Source
+The results from the query needing refinement should be placed into a temporary database to be used for further refinement. This database should be treated as a new temporary source to be queried. Upon completion of the command, this temporary database shall be removed. If the database is already existing and a new command is run, the database shall be overwritten.
```
I don't see any need for a file for this purpose.  Even if SQLite is used, it can be in memory.

JohnMcPMS commented on doc/specs/#1620 - Query Refinement for Disambiguation.md:
```diff
@@ -0,0 +1,275 @@
+The results from the query needing refinement should be placed into a temporary database to be used for further refinement. This database should be treated as a new temporary source to be queried. Upon completion of the command, this temporary database shall be removed. If the database is already existing and a new command is run, the database shall be overwritten.
+
+### Refinement queries
+Input to the package refinement should be treated as a query by default. This should match the behavior of the parent command. Each entry should also be assigned an index number which can be used to reference the package. The index number shall be treated as a fully qualified reference, and entry of an exact match to the number as a query shall take precedent.
```
I would rather we define a smaller scope to the subsequent searches so that it can be implemented via property comparisons rather than needing to use SQLite for full fidelity searching.

JohnMcPMS commented on doc/specs/#1620 - Query Refinement for Disambiguation.md:
```diff
@@ -0,0 +1,275 @@
+### Refinement queries
+Input to the package refinement should be treated as a query by default. This should match the behavior of the parent command. Each entry should also be assigned an index number which can be used to reference the package. The index number shall be treated as a fully qualified reference, and entry of an exact match to the number as a query shall take precedent.
+
+> Example: A user runs `winget install Example` and receives the result:
```
There is an implication that enabling this feature will make the initial query use the same semantics as `search`.  I don't think that you want that.

JohnMcPMS commented on doc/specs/#1620 - Query Refinement for Disambiguation.md:
````diff
@@ -0,0 +1,275 @@
+> ```
+> If the user were to input `'Example 2'`, it would be treated as `winget install -q 'Example 2'` using only the results in the table above as the source
+ 
+Users should be able to refine their query using any of the standard or shared flags. 
````
Again, I would prefer that it be more of a straightforward property search without any need to evaluate complex inputs like this.

JohnMcPMS reviewed (commented):
I would suggest that this feature should be targeted at minor disambiguation, not as a full iterative search mechanism.  The number of results that one can reasonably expect with a single package targeted search is small enough that I think the limit on package count and some of the flows should be reconsidered.
